compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -974,30 +1340,69 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
         }
 
         ty::Dynamic(predicates, _region, kind) => {
-            ty = Ty::new_dynamic(
-                tcx,
-                transform_predicates(tcx, predicates),
-                tcx.lifetimes.re_erased,
-                *kind,
-            );
-        }
-
-        ty::Alias(..) => {
-            ty = transform_ty(
-                tcx,
-                tcx.normalize_erasing_regions(ty::ParamEnv::reveal_all(), ty),
-                options,
-            );
+            if is_fn_trait(tcx, predicates) {
+                // Transform Fn trait objects into function pointers
+                let fn_sig = ty::Binder::bind_with_vars(
+                    fn_trait_sig(tcx, predicates).unwrap(),
+                    List::empty(),
+                );
+                let fn_ptr = Ty::new_fn_ptr(tcx, fn_sig);
+                // Transform fn_sig inputs and output
+                ty = transform_ty(tcx, fn_ptr, options);
+            } else {
+                ty = Ty::new_dynamic(
+                    tcx,
+                    transform_predicates(tcx, predicates, options),
+                    tcx.lifetimes.re_erased,
+                    *kind,
+                );
+            }
         }
 
-        ty::Bound(..) | ty::Error(..) | ty::Infer(..) | ty::Param(..) | ty::Placeholder(..) => {
+        ty::Bound(..)
+        | ty::Error(..)
+        | ty::CoroutineClosure(..)
+        | ty::CoroutineWitness(..)
+        | ty::Infer(..)
+        | ty::Placeholder(..) => {
             bug!("transform_ty: unexpected `{:?}`", ty.kind());
         }
     }
 
     ty
 }
 
+/// Returns a type metadata identifier for the specified drop FnAbi using the Itanium C++ ABI with
```

Comment:
Why is this necessary? In general, I'd like to see every hack be paired with code that motivates why these normalizations must be done.

Reply:
I think I should provide this information initially because it'll help a lot of the answers to the questions below to make sense: `typeid_for_fnabi` is called at two locations, initially when declaring/defining functions and methods, which it has a lot more context/information (i.e., `FnAbi` and `Instance`), and later during code generation at call sites, which it has a lot less contex/information (i.e., an `FnAbi` only, and after type erasure was performed).

In the first call (i.e., when when declaring/defining functions and methods), it encodes and assigns one or more labels to a function or method (these labels are used later by the LowerTypeTests LLVM pass to aggregate functions in groups derived by these labels).

In the second call, it it encodes a given FnAbi (i.e., without an Instance, and after type erasure), and uses this computed label to test if the given pointer/function is member of the group derived from this computed label. It means that in the first call (where we assign labels) we can only include at most as much information that would be available at the call sites, otherwise the labels won't match.

Given this information, this is necessary because the DropGlue generated by the compiler will have indirect calls to a given type drop function. However, Rust allows for types to be erased to any trait object and retains the drop function for the original type, which means at the indirect call sites in theDropGlue generated by the compiler, when `typeid_for_fnabi` is called a second time, it will only have a `FnAbi` after type erasure was performed, and it could be an `FnAbi` to a call on any arbitrary trait object, so we normalize them to a synthesized `Drop` trait object, both on declaration/definition, and the call site, so they are all labelled the same and match, allowing the test to pass and CFI to work with drop.